File: core/src/main/java/org/elkd/core/log/LogInvoker.java
package org.elkd.core.log;

import com.google.common.base.Preconditions;

import javax.annotation.Nonnull;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class LogInvoker<E> implements Log<E> {
  private final Set<LogChangeListener<E>> mListeners = new HashSet<>();
  private final Log<E> mLog;

  public LogInvoker(@Nonnull final Log<E> log) {
    mLog = Preconditions.checkNotNull(log, "log");
  }

  @Override
  public long append(final E entry) {
    final long index = mLog.append(entry);
    onAppend(entry);
    return index;
  }

  @Override
  public long append(final long index, final E entry) {
    final long append = mLog.append(index, entry);
    onAppend(entry);
    return append;
  }

  @Override
  public E read(final long index) {
    return mLog.read(index);
  }

  @Override
  public List<E> read(final long from, final long to) {
    return mLog.read(from, to);
  }

  @Override
  public CommitResult<E> commit(final long index) {
    final CommitResult<E> result = mLog.commit(index);
    onCommit(result.getCommitted());
    return result;
  }

  @Override
  public void revert(final long index) {
    mLog.revert(index);
  }

  @Override
  public long getCommitIndex() {
    return mLog.getCommitIndex();
  }

  @Override
  public long getLastIndex() {
    return mLog.getLastIndex();
  }

  public void registerListener(@Nonnull final LogChangeListener<E> listener) {
    Preconditions.checkNotNull(listener, "listener");
    mListeners.add(listener);
  }

  public void deregisterListener(@Nonnull final LogChangeListener<E> listener) {
    Preconditions.checkNotNull(listener, "listener");
    mListeners.remove(listener);
  }

  private void onCommit(final List<E> entries) {
    for (final LogChangeListener<E> listener : mListeners) {
      entries.forEach(listener::onCommit);
    }
  }

  private void onAppend(final E entry) {
    for (final LogChangeListener<E> listener : mListeners) {
      listener.onAppend(entry);
    }
  }
}

File: core/src/main/java/org/elkd/core/log/LogInvoker.kt
package org.elkd.core.log

import org.elkd.shared.annotations.Mockable
import java.util.HashSet

@Mockable
class LogInvoker<E : LogEntry>(private val log: Log<E>) : Log<E> {
  private val listeners = HashSet<LogChangeListener<E>>()

  override val commitIndex: Long
    get() = log.commitIndex

  override val lastIndex: Long
    get() = log.lastIndex

  override val lastEntry: E
    get() = log.lastEntry

  override fun append(entry: E): Long {
    val index = log.append(entry)
    onAppend(index, entry)
    return index
  }

  override fun read(index: Long): E? {
    return log.read(index)
  }

  override fun read(from: Long, to: Long): List<E> {
    return log.read(from, to)
  }

  override fun append(index: Long, entry: E): Long {
    val append = log.append(index, entry)
    onAppend(append, entry)
    return append
  }

  override fun commit(index: Long): CommitResult<E> {
    val result = log.commit(index)
    result.committed.forEachIndexed { i, e -> onCommit((result.commitIndex - (result.committed.size - 1)) + i, e) }
    return result
  }

  override fun revert(index: Long) {
    log.revert(index)
  }

  internal fun registerListener(listener: LogChangeListener<E>) {
    listeners.add(listener)
  }

  internal fun deregisterListener(listener: LogChangeListener<E>) {
    listeners.remove(listener)
  }

  private fun onCommit(index: Long, entry: E) {
    listeners.forEach { it.onCommit(index, entry) }
  }

  private fun onAppend(index: Long, entry: E) {
    listeners.forEach { it.onAppend(index, entry) }
  }

  override fun toString(): String {
    return "LogInvoker(log=$log)"
  }
}
